nio/client/async_client.py

```diff
@@ -903,25 +906,119 @@ async def login_raw(
 
         return await self._send(LoginResponse, method, path, data)
 
-    async def register(self, username, password, device_name=""):
-        """Register with homeserver.
+    async def register_interactive(
+        self,
+        username: str,
+        password: str,
+        auth_dict: Dict[str, Any],
```

Comment:
I'm very tempted to say that `auth_dict` (here as well as in `api.py`) should be given its own `TypedDict` (which is interchangable/backwards compatible with a normal `dict`) for better introspection of the expected schema. But I'm 10% unsure, so if you have a compelling reason why it's not worth it this may already be fine

Reply:
I agree; however, the only required thing in the auth dictionary is the `session` token. The rest are optional fields that vary depending on the flow the user is attempting to complete. One sort of exception is `type`, which according to the spec, "may be omitted if `session` is given, and the client is reissuing a request which it believes has been completed out-of-band." 

The Matrix spec outlines the Authentication Data (`auth_dict`) in the request description here:
https://spec.matrix.org/latest/client-server-api/#post_matrixclientv3register

I'm still open to the idea, though! If we were to add a type for this, where should it go? We could also set up [login_raw](https://github.com/justin-russell/matrix-nio/blob/registration-tokens/nio/client/async_client.py#L874) in a similar way too!

Reply:
Yeah I'm thinking using `TypedDict` is still viable/useful, but probably would work better with a more pervasive usage of it rather than just here, opportunistically: lgtm as is!